### ISSUE TITLE
dict_free_contents ignores DI_FLAGS_ALLOC absense

### ISSUE THREAD

**Issue by ZyX-I** [vim/vim#1507]
At vim-8.0.348 dict_free_contents looks like

```C
dict_free_contents(dict_T *d)
{
    int         todo;
    hashitem_T  *hi;
    dictitem_T  *di;

    /* Lock the hashtab, we don't want it to resize while freeing items. */
    hash_lock(&d->dv_hashtab);
    todo = (int)d->dv_hashtab.ht_used;
    for (hi = d->dv_hashtab.ht_array; todo > 0; ++hi)
    {
        if (!HASHITEM_EMPTY(hi))
        {
            /* Remove the item before deleting it, just in case there is
             * something recursive causing trouble. */
            di = HI2DI(hi);
            hash_remove(&d->dv_hashtab, hi);
            clear_tv(&di->di_tv);
            vim_free(di);
```

note this part: if dictionary item is not actually allocated it will try to free it.

```C
            --todo;
        }
    }
    hash_clear(&d->dv_hashtab);
}
```

The problem came up when I got an idea: why allocate a dictionary item and bother with access to b_changedtick through pointer (even though I do this only for writes unlike in patch 334) while I can add `dictitem_T` (well, not exactly it, but a compatible type) directly to `buf_T` structure and have it all allocated in one call?

Patch is dead simple: replace `clear_tv(&di->di_tv);` and `vim_free(di);` with `dictitem_free(di);`, this function does exactly the same thing except that it checks for DI_FLAGS_ALLOC first.

**Comment by ZyX-I** [comment]
Though I would suggest to also put b:changedtick dictionary item inside buf_T.

**Comment by brammool** [comment]
Nikolai Pavlov wrote:

> Though I would suggest to also put b:changedtick dictionary item
> inside buf_T.

You mean, to avoid the need to allocate it?

-- 
When a fly lands on the ceiling, does it do a half roll or
a half loop?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ZyX-I** [comment]
Both to avoid the need to allocate it and to avoid the need to use indirect access to changedtick through a pointer.

**Comment by brammool** [comment]
Nikolai Pavlov wrote:

> At vim-8.0.348 dict_free_contents looks like
> 
> ```C
> dict_free_contents(dict_T *d)
> {
>     int         todo;
>     hashitem_T  *hi;
>     dictitem_T  *di;
> 
>     /* Lock the hashtab, we don't want it to resize while freeing items. */
>     hash_lock(&d->dv_hashtab);
>     todo = (int)d->dv_hashtab.ht_used;
>     for (hi = d->dv_hashtab.ht_array; todo > 0; ++hi)
>     {
>         if (!HASHITEM_EMPTY(hi))
>         {
>             /* Remove the item before deleting it, just in case there is
>              * something recursive causing trouble. */
>             di = HI2DI(hi);
>             hash_remove(&d->dv_hashtab, hi);
>             clear_tv(&di->di_tv);
>             vim_free(di);
> ```
> 
> note this part: if dictionary item is not actually allocated it will try to free it.
> 
> ```C
>             --todo;
>         }
>     }
>     hash_clear(&d->dv_hashtab);
> }
> ```
> 
> The problem came up when I got an idea: why allocate a dictionary item and bother with access to b_changedtick through pointer (even though I do this only for writes unlike in patch 334) while I can add `dictitem_T` (well, not exactly it, but a compatible type) directly to `buf_T` structure and have it all allocated in one call?
> 
> Patch is dead simple: replace `clear_tv(&di->di_tv);` and
> `vim_free(di);` with `dictitem_free(di);`, this function does exactly
> the same thing except that it checks for DI_FLAGS_ALLOC first.

Thanks for the idea.  It's not a big win, but avoids an allocation (that
could fail) and it's a bit cleaner.

-- 
hundred-and-one symptoms of being an internet addict:
21. Your dog has its own home page.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
Fixed by [8.0.0365](https://github.com/vim/vim/releases/tag/v8.0.0365)
